#### Naive and Compensated Summation

<a href="https://colab.research.google.com/github/googlecolab/colabtools/blob/master/notebooks/colab-github-demo.ipynb">
  <img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/>
</a>


In [1]:
#Anderson Tan

import time
import math

#defining function for naive summation using its algorithm
def naive_sum(N): 
    sum, value = 0, math.sqrt(13)
    for i in range(N):
        sum = sum + value
    return sum

#defining function for compensated summation using its algorithm
def comp_sum(N):
    s, value, e = 0, math.sqrt(13), 0
    for i in range(N):
        temp = s
        y = value + e
        s = temp + y
        e = (temp - s) + y
    return s + e

#for every different case of N
for N in [10000, 100000, 1000000, 10000000, 100000000]: 
    #calculate the exact sum using different N cases
    exact_sum = N * math.sqrt(13)  
    print('If N =', N,',', 'we get the following:''\n')
    print('Exact Sum:', exact_sum, '\n')
    
    #naive summation
    start_time = time.time() #start time
    naive = naive_sum(N) #defined in function above
    end_time = time.time() #end time
    abs_error = abs(naive-exact_sum) #formula for absolute error
    rel_error = abs((naive-exact_sum)/exact_sum) #formula for relative error
    elapsed_time = end_time-start_time #formula for elapsed time
    
    print('Naive Sum:', naive)
    print('Absolute Error:', abs_error)
    print('Relative Error:', rel_error)
    print('Elapsed Time:', elapsed_time, 'seconds' '\n')
    
    #compensated summation
    start_time = time.time() #start time
    comp = comp_sum(N) #defined in function above
    end_time = time.time() #end time
    abs_error = abs(comp-exact_sum) #formula for absolute error
    rel_error = abs((comp-exact_sum)/exact_sum) #formula for relative error
    elapsed_time = end_time-start_time #formula for elapsed time
    
    print('Compensated Sum:', comp)
    print('Absolute Error:', abs_error)
    print('Relative Error:', rel_error)
    print('Elapsed Time:', elapsed_time, 'seconds' '\n')
    print('-------------------------------------', '\n')

If N = 10000 , we get the following:

Exact Sum: 36055.512754639894 

Naive Sum: 36055.51275464152
Absolute Error: 1.622538547962904e-09
Relative Error: 4.5001122546901057e-14
Elapsed Time: 0.0019943714141845703 seconds

Compensated Sum: 36055.512754639894
Absolute Error: 0.0
Relative Error: 0.0
Elapsed Time: 0.007977962493896484 seconds

------------------------------------- 

If N = 100000 , we get the following:

Exact Sum: 360555.1275463989 

Naive Sum: 360555.12754659855
Absolute Error: 1.996522769331932e-07
Relative Error: 5.537357859582804e-13
Elapsed Time: 0.01895761489868164 seconds

Compensated Sum: 360555.1275463989
Absolute Error: 0.0
Relative Error: 0.0
Elapsed Time: 0.032910823822021484 seconds

------------------------------------- 

If N = 1000000 , we get the following:

Exact Sum: 3605551.275463989 

Naive Sum: 3605551.2754314356
Absolute Error: 3.2553449273109436e-05
Relative Error: 9.028702349800925e-12
Elapsed Time: 0.08477282524108887 seconds

Compensated Sum: 360

||||Table for Naive Summation|||
      |:-|:-|:-|:-|:-|:-|
|N|Exact Sum|Naive Sum|Absolute Error|Relative Error|Elapsed Time (s)| 
|10^4|36055.512754639894|36055.5127546415200|1.6225385479629040e-09|4.5001122546901057e-14|0.0019943714141845703|
|10^5|360555.12754639894|360555.127546598550|1.9965227693319320e-07|5.5373578595828040e-13|0.0189576148986816400|
|10^6|3605551.2754639894|3605551.27543143560|3.2553449273109436e-05|9.0287023498009250e-12|0.0847728252410888700|
|10^7|36055512.754639894|36055512.7588156500|0.00417575985193252600|1.1581474046282028e-10|0.9843652248382568000|
|10^8|360555127.54639894|3605555128.05700934|0.51061040163040160000|1.4161784498951340e-09|7.0711228847503660000|

||||Table for Compensated Summation|||
      |:-|:-|:-|:-|:-|:-|
|N|Exact Sum|Compensated Sum|Absolute Error|Relative Error|Elapsed Time (s)| 
|10^4|36055.512754639894|36055.512754639894|0.0|0.0|0.0079779624938964840|
|10^5|360555.12754639894|360555.12754639894|0.0|0.0|0.0329108238220214840|
|10^6|3605551.2754639894|3605551.2754639894|0.0|0.0|0.2642917633056640600|
|10^7|36055512.754639894|36055512.754639894|0.0|0.0|2.1253137588500977000|
|10^8|360555127.54639894|360555127.54639894|0.0|0.0|22.221544265747070000|

<html>
   <head>
      <title>HTML Font</title>
   </head>

   <body>
      <h4>Explanation</h4>
      <p style = "font-family:georgia,garamond,serif;font-size:12px;font-style:">
         In this program, I calculated SN using two methods, naive and compensated summation. Different trials were performed where N varied between 10^4, 10^5, 10^6, 10^7 and 10^8. I compared these results, for both methods, with the exact sum calculated by N*math.sqrt(13). In addition to this, the time of each summation were also generated by the given code in the assignment. All of the results were stored in the two tables shown above. 
      <p style = "font-family:georgia,garamond,serif;font-size:12px;font-style:"> 
          According to the results, the naive summation method is faster than the compensated summation method in terms of speed. This can be shown through the comparison of the elapsed time from both tables. However, we can also observe that naive summation is actually less accurate than compensated summation. There are no absolute and relative errors of compensated summation compared to the errors of naive summation, although the errors for naive is still a very small number. These errors from the program are, in fact, within the error bounds. Therefore, the errors presented in this project are consistent, as these values are changing in small increments for naive and unchanging for compensated summation. 
      </p>
      <p style = "font-family:georgia,garamond,serif;font-size:12px;font-style:">
          Note that the elapsed time will change once the program is executed again.
      </p>
   </body>

</html>

                                            Meme for one arbitrary brownie point ~ hehe

![alt text](meme.jpeg "MEME")